src/platform/ESP32/nimble/BLEManagerImpl.cpp

```diff
@@ -1040,33 +1040,32 @@ CHIP_ERROR BLEManagerImpl::StartAdvertising(void)
 
     {
         err = ble_gap_adv_start(own_addr_type, NULL, BLE_HS_FOREVER, &adv_params, ble_svr_gap_event, NULL);
-        if (err != CHIP_NO_ERROR)
+        if (err == BLE_HS_EALREADY)
         {
-            ChipLogError(DeviceLayer, "ble_gap_adv_start() failed: %s", ErrorStr(err));
-            ExitNow();
+            /* This error code indicates that the advertising is already active. Stop and restart with the new parameters */
+            ChipLogProgress(DeviceLayer, "Device already advertising, stop active advertisement and restart");
+            err = ble_gap_adv_stop();
+            if (err != 0)
+            {
+                ChipLogError(DeviceLayer, "ble_gap_adv_stop() failed: %s, cannot restart", ErrorStr(err));
+                return CHIP_ERROR_INTERNAL;
+            }
+            else
+            {
+                err = ble_gap_adv_start(own_addr_type, NULL, BLE_HS_FOREVER, &adv_params, ble_svr_gap_event, NULL);
+            }
         }
-    }
 
-exit:
-    if (err == BLE_HS_EALREADY && connectable)
```

Comment:
So why was this not working?  Is `connectable` false in the cases when it failed?

Reply:
A call was made to `StartAdvertising` with updated advertisement data, when the non-connectable advertising was active and it returned error from this function (as the `if` check had the connectable flag). As a result, in `DriveBLEState` the service mode was set to `kCHIPoBLEServiceMode_Disabled` and advertisement was stopped.